test/jdk/java/nio/channels/DatagramChannel/Connect.java

```diff
@@ -43,89 +47,90 @@ public static void main(String[] args) throws Exception {
-        readerThread.join();
-
-        reader.throwException();
-        writer.throwException();
+    static void invoke(ExecutorService e, Runnable reader, Runnable writer) throws CompletionException {
+        CompletableFuture<Void> f1 = CompletableFuture.runAsync(writer, e);
+        CompletableFuture<Void> f2 = CompletableFuture.runAsync(reader, e);
+        wait(f1, f2);
     }
 
-    public interface Sprintable extends Runnable {
-        public void throwException() throws Exception;
+
+    // This method waits for either one of the given futures to complete exceptionally
+    // or for all of the given futures to complete successfully.
+    private static void wait(CompletableFuture<?>... futures) throws CompletionException {
+        CompletableFuture<?> future = CompletableFuture.allOf(futures);
+        Stream.of(futures)
+                .forEach(f -> f.exceptionally(ex -> {
+                    future.completeExceptionally(ex);
+                    return null;
+                }));
+        future.join();
     }
 
-    public static class Actor implements Sprintable {
+    public static class Actor implements AutoCloseable, Runnable {
         final int port;
-        Exception e = null;
+        final DatagramChannel dc;
 
-        Actor(int port) {
+        Actor(int port) throws IOException {
             this.port = port;
-        }
-
-        public void throwException() throws Exception {
-            if (e != null)
-                throw e;
+            dc = DatagramChannel.open();
         }
 
         public void run() {
             try {
-                DatagramChannel dc = DatagramChannel.open();
-
-                // Send a message
                 ByteBuffer bb = ByteBuffer.allocateDirect(256);
                 bb.put("hello".getBytes());
                 bb.flip();
-                InetAddress address = InetAddress.getLocalHost();
-                if (address.isLoopbackAddress()) {
-                    address = InetAddress.getLoopbackAddress();
-                }
+                InetAddress address = InetAddress.getLoopbackAddress();
                 InetSocketAddress isa = new InetSocketAddress(address, port);
                 dc.connect(isa);
+
+                // Send a message
+                log.println("Actor attempting to write to Reactor at " + isa.toString());
                 dc.write(bb);
 
                 // Try to send to some other address
-                address = InetAddress.getLocalHost();
-                InetSocketAddress bogus = new InetSocketAddress(address, 3333);
                 try {
-                    dc.send(bb, bogus);
-                    throw new RuntimeException("Allowed bogus send while connected");
+                    InetSocketAddress otherAddress = new InetSocketAddress(address, (port == 3333 ? 3332 : 3333));
+                    log.println("Testing if Actor throws already connected exception when attempting to send to " + otherAddress.toString());
```

Comment:
Would you mind trimming down this very long line as it makes it too hard to do side-by-side diffs.